### PR TITLE
late materialization with prefix in index and/or early pruning

### PR DESCRIPTION
**Pull request by iurii-i-popov**
Late materialization

Added case when index contains a prefix.
Supported early pruning when index contains necessary fields.

https://github.com/arangodb/backlog/issues/680

http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/9451/

### CODE REVIEW

**Comment by iurii-i-popov** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/9452/

**Comment by iurii-i-popov** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/9453/

File: arangod/Aql/IndexExecutor.cpp [hunk 1]
```diff
@@ -88,8 +89,44 @@ IndexIterator::DocumentCallback getCallback(DocumentProducingFunctionContext& co
...
+        velocypack::Slice slice;
+      };
+      filterContext fc{outNonMaterializedIndVars, slice};
+
+      auto getValue = [](void const* ctx, Variable const* var, bool doCopy) {
+        TRI_ASSERT(ctx && var);
+        auto const& fc = *reinterpret_cast<filterContext const*>(ctx);
+        auto it = fc.outNonMaterializedIndVars.second.find(var);
+        TRI_ASSERT(it != fc.outNonMaterializedIndVars.second.cend());
+        velocypack::Slice s;
```
Comment (Dronplane) [inline arangod/Aql/IndexExecutor.cpp:112]: Maybe return empty AqlValue there if not found (to play safe in release build)?

File: arangod/Aql/IndexNodeOptimizerRules.cpp [hunk 2]
```diff
@@ -122,20 +125,27 @@ void arangodb::aql::lateDocumentMaterializationRule(Optimizer* opt,
...
+      auto commonIndexId = IndexId::none();  // use one index only
       if (indexNode->hasFilter()) {
-        // IndexNode has an early pruning filter. In this case we cannot perform the optimization
-        continue;
+        auto* filter = indexNode->filter();
+        arangodb::containers::HashSet<Variable const*> currentUsedVars;
+        Ast::getReferencedVariables(filter->node(), currentUsedVars);
+        if (currentUsedVars.find(var) != currentUsedVars.end() &&
+            !processCalculationNode(indexNode, nullptr, filter, nodesToChange, commonIndexId)) {
+          // IndexNode has an early pruning filter which references not to index variables.
```
Comment (Dronplane) [inline arangod/Aql/IndexNodeOptimizerRules.cpp:137]:
```suggestion
          // IndexNode has an early pruning filter which references variables not stored in index.
```

File: arangod/Aql/IndexNodeOptimizerRules.cpp [hunk 3]
```diff
@@ -122,20 +125,27 @@ void arangodb::aql::lateDocumentMaterializationRule(Optimizer* opt,
...
+          // In this case we cannot perform the optimization
+          continue;
+        }
       }
-      auto current = limitNode->getFirstDependency();
+      auto* current = limitNode->getFirstDependency();
```
Comment (Dronplane) [inline arangod/Aql/IndexNodeOptimizerRules.cpp:142]: check current against nullptr ?

Comment (iurii-i-popov) [reply arangod/Aql/IndexNodeOptimizerRules.cpp:144]: I have just replaced other auto with auto*. There are many places. But nullptr cannot be there. I added TRI_ASSERT().

**Comment by iurii-i-popov** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/9454/

**Comment by iurii-i-popov** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/9465/

**Comment by iurii-i-popov** [comment]
IndexId changed.

http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/9468/

**Pull request merged**